File: cmd/credit.go
package cmd

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/dalang-io/dalang-cli/internal/api"
)

func cmdCredit(args []string) error {
	if len(args) == 0 {
		return creditBalance()
	}

	switch args[0] {
	case "history":
		return creditHistory()
	case "add":
		if len(args) < 2 {
			printUsage("Usage: dalang credit add <amount>")
			return fmt.Errorf("missing amount")
		}
		return creditAdd(args[1])
	case "--help", "-h":
		printCreditHelp()
		return nil
	default:
		// Check if it's a number (shorthand for add)
		if _, err := strconv.Atoi(args[0]); err == nil {
			return creditAdd(args[0])
		}
		return fmt.Errorf("unknown credit subcommand: %s", args[0])
	}
}

func creditBalance() error {
	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}
	client.Verbose = VerboseOutput

	resp, err := client.Get("/credits/balance")
	if err != nil {
		return fmt.Errorf("failed to fetch balance: %w", err)
	}

	var balanceResp api.BalanceResponse
	if err := json.Unmarshal(resp, &balanceResp); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if jsonOutput {
		data, _ := json.MarshalIndent(balanceResp.Data, "", "  ")
		fmt.Println(string(data))
		return nil
	}

	fmt.Printf("\n%sCredit Balance%s\n", colorBold, colorReset)
	fmt.Println(strings.Repeat("─", 40))
	fmt.Printf("  Balance:     %s%s%s\n", colorGreen+colorBold, formatIDR(balanceResp.Data.Balance), colorReset)
	fmt.Printf("  Total Topup: %s\n", formatIDR(balanceResp.Data.TotalTopup))
	fmt.Printf("  Total Spent: %s\n", formatIDR(balanceResp.Data.TotalSpent))
	if balanceResp.Data.TotalCommission > 0 {
		fmt.Printf("  Commission:  %s%s%s\n", colorCyan, formatIDR(balanceResp.Data.TotalCommission), colorReset)
	}
	fmt.Println()

	return nil
}

func creditHistory() error {
	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}
	client.Verbose = VerboseOutput

	resp, err := client.Get("/credits/transactions?page=1&limit=25")
	if err != nil {
		return fmt.Errorf("failed to fetch transactions: %w", err)
	}

	var txResp api.TransactionsResponse
	if err := json.Unmarshal(resp, &txResp); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if jsonOutput {
		data, _ := json.MarshalIndent(txResp.Data, "", "  ")
		fmt.Println(string(data))
		return nil
	}

	if len(txResp.Data.Transactions) == 0 {
		printInfo("No transactions found")
		return nil
	}

	fmt.Printf("\n%sRecent Transactions%s (showing %d of %d)\n", colorBold, colorReset, len(txResp.Data.Transactions), txResp.Data.Total)
	fmt.Println(strings.Repeat("─", 70))
	fmt.Printf("%-12s %-10s %15s %15s  %s\n", "DATE", "TYPE", "AMOUNT", "BALANCE", "DESCRIPTION")
	fmt.Println(strings.Repeat("─", 70))

	for _, tx := range txResp.Data.Transactions {
		date := formatDate(tx.CreatedAt)
		typeColor := colorReset
		amountStr := formatIDR(tx.Amount)

		switch tx.Type {
		case "topup":
			typeColor = colorGreen
			amountStr = "+" + amountStr
		case "spend":
			typeColor = colorRed
			amountStr = "-" + amountStr
		case "commission":
			typeColor = colorCyan
			amountStr = "+" + amountStr
		case "refund":
			typeColor = colorYellow
			amountStr = "+" + amountStr
		}

		desc := tx.Description
		if len(desc) > 25 {
			desc = desc[:22] + "..."
		}

		fmt.Printf("%-12s %s%-10s%s %15s %15s  %s\n",
			date,
			typeColor, tx.Type, colorReset,
			amountStr,
			formatIDR(tx.BalanceAfter),
			desc,
		)
	}
	fmt.Println()

	return nil
}

func creditAdd(amountStr string) error {
	amount, err := strconv.Atoi(amountStr)
	if err != nil {
		return fmt.Errorf("invalid amount: %s", amountStr)
	}

	// Convert to actual amount (input is in thousands)
	actualAmount := amount * 1000

	if actualAmount < 50000 {
		return fmt.Errorf("minimum topup is 50K IDR (enter 50 or higher)")
	}

	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}

	printInfo("Creating topup for %s...", formatIDR(int64(actualAmount)))

	resp, err := client.Post("/credits/topup", map[string]interface{}{
		"amount": actualAmount,
	})
	if err != nil {
		return fmt.Errorf("failed to create topup: %w", err)
	}

	var topupResp api.TopupResponse
	if err := json.Unmarshal(resp, &topupResp); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if jsonOutput {
		data, _ := json.MarshalIndent(topupResp.Data, "", "  ")
		fmt.Println(string(data))
		return nil
	}

	fmt.Println()
	printSuccess("Topup invoice created!")
	fmt.Printf("\n  %sAmount:%s %s\n", colorBold, colorReset, formatIDR(int64(topupResp.Data.Amount)))
	fmt.Printf("  %sPayment URL:%s\n", colorBold, colorReset)
	fmt.Printf("  %s%s%s\n\n", colorCyan, topupResp.Data.InvoiceURL, colorReset)
	printInfo("Open the URL above to complete payment")

	return nil
}

func formatIDR(amount int64) string {
	prefix := "Rp "
	if amount < 0 {
		prefix = "Rp -"
		amount = -amount
	}

	// Format with thousand separators
	str := fmt.Sprintf("%d", amount)
	n := len(str)
	if n <= 3 {
		return prefix + str
	}

	var result strings.Builder
	result.WriteString(prefix)

	remainder := n % 3
	if remainder > 0 {
		result.WriteString(str[:remainder])
		if n > remainder {
			result.WriteString(".")
		}
	}

	for i := remainder; i < n; i += 3 {
		result.WriteString(str[i : i+3])
		if i+3 < n {
			result.WriteString(".")
		}
	}

	return result.String()
}

func formatDate(isoDate string) string {
	if len(isoDate) >= 10 {
		return isoDate[:10]
	}
	return isoDate
}

func formatExpiryWithDays(isoDate string) string {
	if isoDate == "" {
		return "N/A"
	}

	// Parse the date
	t, err := time.Parse("2006-01-02T15:04:05Z", isoDate)
	if err != nil {
		// Try alternative format
		t, err = time.Parse("2006-01-02 15:04:05", isoDate)
		if err != nil {
			t, err = time.Parse("2006-01-02", isoDate)
			if err != nil {
				return formatDate(isoDate)
			}
		}
	}

	now := time.Now()
	days := int(t.Sub(now).Hours() / 24)

	dateStr := t.Format("2006-01-02")

	if days < 0 {
		return fmt.Sprintf("%s (%sexpired%s)", dateStr, colorRed, colorReset)
	} else if days == 0 {
		return fmt.Sprintf("%s (%stoday%s)", dateStr, colorYellow, colorReset)
	} else if days == 1 {
		return fmt.Sprintf("%s (%s1 day left%s)", dateStr, colorYellow, colorReset)
	} else if days <= 7 {
		return fmt.Sprintf("%s (%s%d days left%s)", dateStr, colorYellow, days, colorReset)
	} else {
		return fmt.Sprintf("%s (%d days left)", dateStr, days)
	}
}

func printCreditHelp() {
	fmt.Printf(`%sdalang credit%s - Manage credits/wallet

%sUSAGE:%s
    dalang credit            Show current balance
    dalang credit history    Show recent transactions
    dalang credit add <N>    Top up N thousand IDR (e.g., 50 = 50K)

%sDESCRIPTION:%s
    Manage your Dalang credits balance. Credits can be used to pay
    for VPS, containers, and app hosting services.

%sEXAMPLES:%s
    dalang credit            # Check balance
    dalang credit history    # View last 25 transactions
    dalang credit add 100    # Top up 100K IDR
    dalang credit add 500    # Top up 500K IDR

%sNOTE:%s
    Minimum topup is 50K IDR. Credits expire 12 months after topup.
`,
		colorCyan, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
	)
}

File: src/commands/credit.rs
use crate::api::{self, Client};
use crate::output::*;

pub fn cmd_credit(args: &[String]) -> Result<(), String> {
    if args.is_empty() {
        return credit_balance();
    }

    match args[0].as_str() {
        "history" => credit_history(),
        "add" => {
            if args.len() < 2 {
                print_error("Usage: dalang credit add <amount>");
                return Err("missing amount".to_string());
            }
            credit_add(&args[1])
        }
        "--help" | "-h" => {
            crate::commands::help::print_credit_help();
            Ok(())
        }
        other => {
            // Check if it's a number (shorthand for add)
            if other.parse::<i64>().is_ok() {
                return credit_add(other);
            }
            print_error(&format!("Unknown credit subcommand: {}", other));
            Err(format!("unknown subcommand: {}", other))
        }
    }
}

fn credit_balance() -> Result<(), String> {
    let client = Client::new_authenticated()?;

    let resp = client
        .get("/credits/balance")
        .map_err(|e| format!("failed to fetch balance: {}", e))?;

    let balance_resp: api::BalanceResponse =
        serde_json::from_slice(&resp).map_err(|e| format!("failed to parse response: {}", e))?;

    if json_output() {
        println!(
            "{}",
            serde_json::to_string_pretty(&balance_resp.data).unwrap()
        );
        return Ok(());
    }

    println!();
    println!("{}Credit Balance{}", BOLD, RESET);
    println!("{}", "\u{2500}".repeat(40));
    println!(
        "  Balance:     {}{}{}{}",
        GREEN,
        BOLD,
        format_idr(balance_resp.data.balance),
        RESET
    );
    println!(
        "  Total Topup: {}",
        format_idr(balance_resp.data.total_topup)
    );
    println!(
        "  Total Spent: {}",
        format_idr(balance_resp.data.total_spent)
    );
    if balance_resp.data.total_commission > 0 {
        println!(
            "  Commission:  {}{}{}",
            CYAN,
            format_idr(balance_resp.data.total_commission),
            RESET
        );
    }
    println!();

    Ok(())
}

fn credit_history() -> Result<(), String> {
    let client = Client::new_authenticated()?;

    let resp = client
        .get("/credits/transactions?page=1&limit=25")
        .map_err(|e| format!("failed to fetch transactions: {}", e))?;

    let tx_resp: api::TransactionsResponse =
        serde_json::from_slice(&resp).map_err(|e| format!("failed to parse response: {}", e))?;

    if json_output() {
        println!(
            "{}",
            serde_json::to_string_pretty(&tx_resp.data).unwrap()
        );
        return Ok(());
    }

    if tx_resp.data.transactions.is_empty() {
        print_info("No transactions found");
        return Ok(());
    }

    println!(
        "\n{}Recent Transactions{} (showing {} of {})",
        BOLD,
        RESET,
        tx_resp.data.transactions.len(),
        tx_resp.data.total
    );
    println!("{}", "\u{2500}".repeat(70));
    println!(
        "{:<12} {:<10} {:>15} {:>15}  {}",
        "DATE", "TYPE", "AMOUNT", "BALANCE", "DESCRIPTION"
    );
    println!("{}", "\u{2500}".repeat(70));

    for tx in &tx_resp.data.transactions {
        let date = format_date(&tx.created_at);
        let (type_color, amount_str) = match tx.tx_type.as_str() {
            "topup" => (GREEN, format!("+{}", format_idr(tx.amount))),
            "spend" => {
                let abs_amount = if tx.amount < 0 { -tx.amount } else { tx.amount };
                (RED, format!("-{}", format_idr(abs_amount)))
            }
            "commission" => (CYAN, format!("+{}", format_idr(tx.amount))),
            "refund" => (YELLOW, format!("+{}", format_idr(tx.amount))),
            _ => (RESET, format_idr(tx.amount)),
        };

        let desc = if tx.description.len() > 25 {
            format!("{}...", &tx.description[..22])
        } else {
            tx.description.clone()
        };

        println!(
            "{:<12} {}{:<10}{} {:>15} {:>15}  {}",
            date,
            type_color,
            tx.tx_type,
            RESET,
            amount_str,
            format_idr(tx.balance_after),
            desc,
        );
    }
    println!();

    Ok(())
}

fn credit_add(amount_str: &str) -> Result<(), String> {
    let amount: i64 = amount_str
        .parse()
        .map_err(|_| format!("invalid amount: {}", amount_str))?;

    let actual_amount = amount * 1000;

    if actual_amount < 50000 {
        print_error("Minimum topup is 50K IDR (enter 50 or higher)");
        return Err("minimum topup is 50K".to_string());
    }

    let client = Client::new_authenticated()?;

    print_info(&format!("Creating topup for {}...", format_idr(actual_amount)));

    let body = serde_json::json!({"amount": actual_amount});
    let resp = client
        .post("/credits/topup", Some(&body))
        .map_err(|e| format!("failed to create topup: {}", e))?;

    let topup_resp: api::TopupResponse =
        serde_json::from_slice(&resp).map_err(|e| format!("failed to parse response: {}", e))?;

    if json_output() {
        println!(
            "{}",
            serde_json::to_string_pretty(&topup_resp.data).unwrap()
        );
        return Ok(());
    }

    println!();
    print_success("Topup invoice created!");
    println!(
        "\n  {}Amount:{} {}",
        BOLD,
        RESET,
        format_idr(topup_resp.data.amount)
    );
    println!("  {}Payment URL:{}", BOLD, RESET);
    println!("  {}{}{}\n", CYAN, topup_resp.data.invoice_url, RESET);
    print_info("Open the URL above to complete payment");

    Ok(())
}

pub fn format_idr(amount: i64) -> String {
    let s = format!("{}", amount);
    let n = s.len();
    if n <= 3 {
        return format!("Rp {}", s);
    }

    let mut result = String::from("Rp ");
    let remainder = n % 3;
    if remainder > 0 {
        result.push_str(&s[..remainder]);
        if n > remainder {
            result.push('.');
        }
    }

    let mut i = remainder;
    while i < n {
        result.push_str(&s[i..i + 3]);
        if i + 3 < n {
            result.push('.');
        }
        i += 3;
    }

    result
}

pub fn format_date(iso_date: &str) -> String {
    if iso_date.len() >= 10 {
        iso_date[..10].to_string()
    } else {
        iso_date.to_string()
    }
}

pub fn format_expiry_with_days(iso_date: &str) -> String {
    if iso_date.is_empty() {
        return "N/A".to_string();
    }

    let date_str = format_date(iso_date);

    // Try to parse date and calculate days remaining
    // Parse YYYY-MM-DD
    let parts: Vec<&str> = date_str.split('-').collect();
    if parts.len() != 3 {
        return date_str;
    }

    let (year, month, day) = match (
        parts[0].parse::<i64>(),
        parts[1].parse::<i64>(),
        parts[2].parse::<i64>(),
    ) {
        (Ok(y), Ok(m), Ok(d)) => (y, m, d),
        _ => return date_str,
    };

    // Get current date
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64;

    // Approximate: convert expiry to epoch seconds
    let expiry_epoch = date_to_epoch(year, month, day);
    let now_days = now / 86400;
    let expiry_days = expiry_epoch / 86400;
    let days = expiry_days - now_days;

    if days < 0 {
        format!("{} ({}expired{})", date_str, RED, RESET)
    } else if days == 0 {
        format!("{} ({}today{})", date_str, YELLOW, RESET)
    } else if days == 1 {
        format!("{} ({}1 day left{})", date_str, YELLOW, RESET)
    } else if days <= 7 {
        format!("{} ({}{} days left{})", date_str, YELLOW, days, RESET)
    } else {
        format!("{} ({} days left)", date_str, days)
    }
}

fn date_to_epoch(year: i64, month: i64, day: i64) -> i64 {
    // Simple epoch calculation (not accounting for all edge cases, but sufficient)
    let mut y = year;
    let mut m = month;
    if m <= 2 {
        y -= 1;
        m += 12;
    }
    let days = 365 * y + y / 4 - y / 100 + y / 400 + (153 * (m - 3) + 2) / 5 + day - 719469;
    days * 86400
}
